### ISSUE TITLE
Saving

### ISSUE THREAD

**Issue by chrisotoro** [Ro6ertWcislo/AutoESN#7]
Any recommendations on how to best save a model? With other esn libraries, I typically save the relevant hyper-parameters / random seed and retrain but this seems like it might be a little more nuanced. 

Also, I noticed that the to_cuda() call was not used on your Mackey example with the gdesn- is there an issue with using cuda in that case? 

I have really enjoyed reviewing your work. It is excellent. Thank you for being so thorough in your tests.

**Comment by Ro6ertWcislo** [comment]
Hi, thank you for trying it out! 
Unfortunately, there is no feature for saving a model at this moment. I've tried pickling right now, but it doesn't work, because I use a lot of inner functions. I will implement saving a model ASAP - this week for sure, maybe around Thursday/Friday.

As for cuda - for such small series like MackeyGlass there is usually no profit in using GPU - at least from my experience. There is however no reason not to use it if you have one. I reviewed it and fixed some obvious mistakes, so please fetch latest changes from main branch and the example should work with cuda. 
I didnt have enough time to test it against different scenarios, so the fix may not be robust, but I'll go back to it later this week to make sure it really works.
Please let me know if you need anything else, or would like to contribute somehow :)

**Comment by chrisotoro** [comment]
Thank you, I really appreciate it!

Another question: using mg_example.py as a starting point, are you able to successfully input an X w/ a dataset that has an input_size greater than 1? It appears that the check_forward_input in auto_esn/esn/reservoir/cell.py might not be receiving the full dimensions of the input. I'm still looking into it to see if it might be an error on my end. Right now,  I have X following size torch.Size([series length, number of features]) and the input_size = number of features. Should I be doing something different?

**Comment by Ro6ertWcislo** [comment]
Hello again, it works for me :) I created a small snipped in the pull request here: https://github.com/Ro6ertWcislo/AutoESN/pull/9
Please observe the changes and try running it. I duplicated the mackeyglass input, so that now I have two same series as input, so I get a tensor shaped (num_samples, 2) but the result is still one series.
What about your output dimension? If it's greater than one, you should check the "output_dim" parameter. 
Look here: https://github.com/Ro6ertWcislo/AutoESN/pull/10
I created two-dimensional input with two mackey glasses. The second is shifted (-1 for the whole series). And after setting the output_dim=2 I'm getting two different series as the output.
Let me know if this is helpful. 
Btw, I'm not 100% sure if I'm going to make it with saving until tomorrow, but I'll add it by the end of the week for sure

**Comment by Ro6ertWcislo** [comment]
Hi, I added saving - please take a look at https://github.com/Ro6ertWcislo/AutoESN/blob/master/src/auto_esn/examples/save_load_model.py
you just call a torch.save() and that's it :)

You had a really good point with saving only the seed and then recreating the whole model, but unfortunately I had to do it "the hard way" saving everything. That's because of the Initialization object which can be really complex and I'm using a lot of inner functions there, which cannot be pickled. Moreover, I would have to remember not only the seed, but also all the operations performed during initialization and their parameters . This could be done, but it would require a lot of work, so just saving a model with torch utilities is the solution for now :D
Let me know if it works for you and if you have any other questions

**Comment by chrisotoro** [comment]
Thank you, this is fantastic. I will check out the torch.save() addition. It makes sense that it would be difficult to pickle the model. 

Your clarification regarding the multivariate inputs was very helpful. Thank you for creating that example. I know It's not unique to this model but it appears that the accuracy is very sensitive to the scaling of the multivariate inputs.

**Comment by Ro6ertWcislo** [comment]
hm, have you tried using self normalizing activation?
activation = self_normalizing_default(leaky_rate=1.0, spectral_radius=500)
For me it looked like it's more robust against input that you would scale otherwise, especially that the spectral radius in this activation enables you to omit the scaling part, as shown here https://user-images.githubusercontent.com/32894690/113477802-ce242000-9484-11eb-8c35-b3ee36a4b394.png (image at the bottom of readme).

Other things htatthat worked well for me was to normalize the input, remember the shift and spread, then train the model on normalize input and during inferece I would have my result as model_output*spread + shift.
An example is here https://github.com/Ro6ertWcislo/AutoESN/blob/master/src/auto_esn/examples/sunspot.py

Im not 100% sure it would help you, but I guess it's worth trying :)
